### PR TITLE
Re-enable mark-scan backend test coverage and add tests

### PR DESCRIPTION
**Pull request by kshen0**
## Overview
Mostly reviewable commit-by-commit. The last 3-4 commits are clean up and merge conflict resolution.

https://github.com/votingworks/vxsuite/issues/3874

Re-enables test coverage in mark-scan backend. Tests were previously disabled because behavior and architecture were frequently in flux. It's not 100% but it covers the majority of paths that were testable with reasonable effort. Might be worth a second PR later for the harder-to-test paths.

## Demo Video or Screenshot
N/A

## Testing Plan

This PR is only tests.

## Checklist

- [x] I have added [logging](https://github.com/votingworks/vxsuite/tree/main/libs/logging) where appropriate to any new user actions, system updates such as file reads or storage writes, or errors introduced.

### CODE REVIEW

File: apps/mark-scan/backend/test/auth_helpers.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  electionDefinition: ElectionDefinition
+): void {
+  mockOf(auth.getAuthStatus).mockImplementation(() =>
+    Promise.resolve({
+      status: 'logged_in',
+      user: fakePollWorkerUser(electionDefinition),
+      sessionExpiresAt: fakeSessionExpiresAt(),
+    })
+  );
+}
```
Comment (kshen0) [inline apps/mark-scan/backend/test/auth_helpers.ts:35]: These were factored out from the next commit https://github.com/votingworks/vxsuite/pull/4628/commits/73160461edba0c6a6da5e7026a52e940c721dbe6#diff-174d3fd242de548eda5dd587bdaa2bd8fad20ec94d50930f9ec62c285e3ed6daL93

File: apps/mark-scan/backend/src/app.test.ts [hunk 2]
```diff
@@ -225,31 +221,8 @@ async function expectElectionState(expected: Partial<ElectionState>) {
...
-        SystemSettingsSchema
-      ).unsafeUnwrap(),
-    })
-  );
-
-  const writeResult = await apiClient.configureElectionPackageFromUsb();
-  assert(writeResult.isOk());
-
-  usbDrive.removeUsbDrive();
-}
```
Comment (kshen0) [inline apps/mark-scan/backend/src/app.test.ts:248]: This was mostly redundant with `setUpUsbAndConfigureElection`

File: apps/mark-scan/backend/src/app.test.ts [hunk 3]
```diff
@@ -362,20 +331,116 @@ test('printing a ballot increments the printed ballot count', async () => {
...
+  driver.setPaperHandlerStatus({
+    paperInputLeftInnerSensor: true,
+    paperInputRightInnerSensor: true,
+    paperInputLeftOuterSensor: true,
+    paperInputRightOuterSensor: true,
+    paperNotPresent: false,
+    paperOutSensor: false,
+    paperJam: false,
+  });
+}
```
Comment (kshen0) [inline apps/mark-scan/backend/src/app.test.ts:405]: Going to take a second look at this to see if `jest` can be used instead of the mock driver

File: apps/mark-scan/backend/src/custom-paper-handler/cli/state_machine_cli.ts [hunk 4]
```diff
@@ -1,3 +1,4 @@
+/* istanbul ignore next */
```
Comment (kshen0) [inline apps/mark-scan/backend/src/custom-paper-handler/cli/state_machine_cli.ts:1]: This is a dev tool that might be removed soon, so not investing in tests right now.

File: apps/mark-scan/backend/jest.config.js [hunk 5]
```diff
@@ -16,16 +16,13 @@ module.exports = {
...
+    '!test/**/*'
   ],
   coverageThreshold: {
     global: {
-      statements: 100,
-      branches: 100,
-      functions: 100,
-      lines: 100,
+      branches: -16,
+      lines: -10,
```
Comment (kshen0) [inline apps/mark-scan/backend/jest.config.js:25]: Coverage is roughly 89-97% depending on the metric

File: apps/mark-scan/backend/src/server.test.ts [hunk 6]
```diff
@@ -1,25 +1,102 @@
-import { fakeLogger } from '@votingworks/logging';
+import { LogEventId, fakeLogger } from '@votingworks/logging';
 import tmp from 'tmp';
 import { buildMockInsertedSmartCardAuth } from '@votingworks/auth';
+import {
+  BooleanEnvironmentVariableName,
+  getFeatureFlagMock,
+} from '@votingworks/utils';
+import { MockPaperHandlerDriver } from '@votingworks/custom-paper-handler';
+// import { sleep } from '@votingworks/basics';
```
Comment (kshen0) [inline apps/mark-scan/backend/src/server.test.ts]: Cleaned up in https://github.com/votingworks/vxsuite/pull/4628/commits/1fed52f30b1b41daae362e59af77ba38c864bcfc

File: apps/mark-scan/backend/src/server.test.ts [hunk 7]
```diff
@@ -1,25 +1,102 @@
...
+const featureFlagMock = getFeatureFlagMock();
+jest.mock('@votingworks/utils', (): typeof import('@votingworks/utils') => {
+  return {
+    ...jest.requireActual('@votingworks/utils'),
+    isFeatureFlagEnabled: (flag: BooleanEnvironmentVariableName) =>
+      featureFlagMock.isEnabled(flag),
+  };
+});
+
+beforeEach(() => {});
```
Comment (kshen0) [inline apps/mark-scan/backend/src/server.test.ts]: Cleaned up in https://github.com/votingworks/vxsuite/pull/4628/commits/1fed52f30b1b41daae362e59af77ba38c864bcfc

File: apps/mark-scan/backend/src/util/auth.ts [hunk 8]
```diff
@@ -1,3 +1,4 @@
+/* istanbul ignore next */
```
Comment (adghayes) [inline apps/mark-scan/backend/src/util/auth.ts:1]: nit: slight preference for ignoring at the `jest.config` level so we have an inventory of what files we're ignoring

File: apps/mark-scan/backend/src/pat-input/connection_status_reader.ts [hunk 9]
```diff
@@ -44,6 +45,12 @@ export class PatConnectionStatusReader
     return true;
   }
 
+  async close(): Promise<void> {
+    if (this.file) {
+      await this.file.close();
+    }
+  }
+
```
Comment (adghayes) [inline apps/mark-scan/backend/src/pat-input/connection_status_reader.ts:53]: I'm a bit confused in context - why did you add a `close` method here, why wasn't it necessary before?

File: apps/mark-scan/backend/package.json [hunk 10]
```diff
@@ -86,7 +87,8 @@
     "lint-staged": "11.0.0",
     "nodemon": "^2.0.20",
     "sort-package-json": "^1.50.0",
-    "ts-jest": "29.1.1"
+    "ts-jest": "29.1.1",
+    "wait-for-expect": "^3.0.2"
```
Comment (adghayes) [inline apps/mark-scan/backend/package.json:91]: nit: there is a `backendWaitFor` helper in test-utils

**Review by adghayes — APPROVED** [review]
glad you had the opportunity to focus on this and make the mocking and testing more tractable for yourself and future editors. I left mostly nit comments and some questions about the hardware check skip. The latest rebase is broken probably due to testing I added in a follow-up logging PR.

File: apps/mark-scan/backend/src/server.test.ts [hunk 11]
```diff
@@ -1,25 +1,102 @@
...
+  expect(logger.log).toHaveBeenCalledWith(
+    LogEventId.PaperHandlerConnection,
+    'system',
+    {
+      disposition: 'failure',
+    }
+  );
+});
+
+// Something about the feature flag mock (?) makes these tests not terminate cleanly
```
[the remark this answers is not in the document]
Comment (adghayes) [reply apps/mark-scan/backend/src/server.test.ts:59]: looks like it's not uncommented. @kshen0 I'm not sure what it would have to do with the feature flag mock, but you could reset all flags between tests (it doesn't happen automatically, but maybe we could set that up)

File: apps/mark-scan/backend/src/pat-input/connection_status_reader.ts (hunk 9 above)
Comment (kshen0) [reply apps/mark-scan/backend/src/pat-input/connection_status_reader.ts:53]: When trying to fix hanging tests, I was looking for anything that might have an unresolved promise, open file handle, etc. While this didn't fix the test issue, I kept it because it seemed appropriate to provide a `close()` function when we have an `open()` one.

File: apps/mark-scan/backend/src/server.test.ts [hunk 12]
```diff
@@ -1,25 +1,102 @@
...
+    isFeatureFlagEnabled: (flag: BooleanEnvironmentVariableName) =>
+      featureFlagMock.isEnabled(flag),
+  };
+});
+
+beforeEach(() => {});
+
+afterEach(() => {
+  jest.resetAllMocks();
+});
```
[the remark this answers is not in the document]
Comment (kshen0) [reply apps/mark-scan/backend/src/server.test.ts]: I tested and you're right. Originally added this to try to solve the hanging test issue but seems to be unnecessary. [4f5e970](https://github.com/votingworks/vxsuite/pull/4628/commits/4f5e970dadc65296b84d4e213cdbaafb3998de55)

File: apps/mark-scan/backend/src/server.test.ts [hunk 13]
```diff
@@ -1,25 +1,102 @@
...
   });
   expect(server.listening).toBeTruthy();
   server.close();
-  stateMachine.stopMachineService();
+  workspace.reset();
 });
+
+test('resolveDriver returns a mock driver if feature flag is on', async () => {
+  featureFlagMock.enableFeatureFlag(
+    BooleanEnvironmentVariableName.SKIP_PAPER_HANDLER_HARDWARE_CHECK
```
[the remark this answers is not in the document]
Comment (kshen0) [reply apps/mark-scan/backend/src/server.test.ts]: [996594a](https://github.com/votingworks/vxsuite/pull/4628/commits/996594a4683edbae96b2418ef2140e10db9f81be)

File: apps/mark-scan/backend/package.json (hunk 10 above)
Comment (kshen0) [reply apps/mark-scan/backend/package.json]: [d8ecba4](https://github.com/votingworks/vxsuite/pull/4628/commits/d8ecba46a7037fb31710b150a56c37b4a61dfd4b)

File: apps/mark-scan/backend/src/util/auth.ts (hunk 8 above)
Comment (kshen0) [reply apps/mark-scan/backend/src/util/auth.ts]: [ae0bad8](https://github.com/votingworks/vxsuite/pull/4628/commits/ae0bad8230459bc59a5f5ab2338f12a86326b9bc)

File: apps/mark-scan/backend/src/server.ts [hunk 14]
```diff
@@ -34,35 +32,34 @@ export interface StartOptions {
...
   workspace: Workspace;
-  // Allow undefined state machine to fail gracefully if no connection to paper handler
-  stateMachine?: PaperHandlerStateMachine;
 }
 
-async function resolveDriver(
+export async function resolveDriver(
   logger: Logger
 ): Promise<PaperHandlerDriverInterface | undefined> {
   const driver = await getPaperHandlerDriver();
```
[the remark this answers is not in the document]
Comment (kshen0) [reply apps/mark-scan/backend/src/server.ts:46]: [df80a8f](https://github.com/votingworks/vxsuite/pull/4628/commits/df80a8fdee01e009869925686b88533fbea85f57)

File: apps/mark-scan/backend/src/server.ts [hunk 15]
```diff
@@ -34,35 +32,34 @@ export interface StartOptions {
...
     await logger.log(LogEventId.PaperHandlerConnection, 'system', {
       disposition: 'success',
     });
+    return driver;
   }
 
-  return driver;
+  await logger.log(LogEventId.PaperHandlerConnection, 'system', {
+    disposition: 'failure',
+  });
```
[the remark this answers is not in the document]
Comment (kshen0) [reply apps/mark-scan/backend/src/server.ts:57]: [0c1d661](https://github.com/votingworks/vxsuite/pull/4628/commits/0c1d661326ba6afb4590389c5d061d4f24097e14)

File: apps/mark-scan/backend/src/server.test.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (kshen0) [reply apps/mark-scan/backend/src/server.test.ts]: [7024b8e](https://github.com/votingworks/vxsuite/pull/4628/commits/7024b8e922b2c2e0eccd8b2266f8e1e5628c55af)

I did a little digging and it might be more related to the business logic of the flag than the flag mocking. I was able to add back in the mocked auth test but the mocked driver test had the same problem, so I removed it entirely and cleaned up the comments.

**Review by adghayes — APPROVED** [review]
LGTM
